File: src/01_data_structures/07_graph/01_adjacency_list/implementations/ts/adjacency_list.ts
type Graph = Record<string, Array<Vertex>>;

type Vertex = {
  value: string;
  weight: number;
};

export class AdjacencyList {
  graph: Graph = {};

  getGraph(): Graph {
    return this.graph;
  }

  addVertex(vertex: string): AdjacencyList {
    if (!this.graph[vertex]) {
      this.graph[vertex] = [];
    }

    return this;
  }

  removeVertex(vertex: string): AdjacencyList {
    if (!this.graph[vertex]) {
      return this;
    }

    this.graph[vertex].forEach((item) => {
      this.graph[item.value] = this.graph[item.value].filter(
        (item) => item.value !== vertex
      );
    });

    delete this.graph[vertex];

    return this;
  }

  addEdge(
    u: string,
    v: string,
    options: { directed?: boolean; weight?: number } = {}
  ): AdjacencyList {
    if (!this.graph[u] || !this.graph[v]) {
      return this;
    }

    this.graph[u].push({ value: v, weight: options.weight });

    if (!options.directed) {
      this.graph[v].push({ value: u, weight: options.weight });
    }

    return this;
  }

  removeEdge(
    u: string,
    v: string,
    options: { directed?: boolean } = {}
  ): AdjacencyList {
    if (!this.graph[u] || !this.graph[v]) {
      return this;
    }

    this.graph[v] = this.graph[v].filter((item) => item.value !== u);

    if (!options.directed) {
      this.graph[u] = this.graph[u].filter((item) => item.value !== v);
    }

    return this;
  }

  getNeighbors(vertex: string): Array<Vertex> | null {
    if (!this.graph[vertex]) {
      return null;
    }

    return this.graph[vertex];
  }
}

File: src/01_data_structures/07_graph/01_adjacency_list/implementations/go/adjacency_list.go
package adjacency_list

type EdgeOptions struct {
	Directed bool
	Weight   int
}

type Vertex struct {
	Value  string
	Weight int
}

type AdjacencyList struct {
	graph map[string][]Vertex
}

func NewAdjacencyList() *AdjacencyList {
	return &AdjacencyList{
		graph: make(map[string][]Vertex),
	}
}

func (adjacencyList *AdjacencyList) AddVertex(u string) *AdjacencyList {
	if _, exists := adjacencyList.graph[u]; !exists {
		adjacencyList.graph[u] = []Vertex{}
	}

	return adjacencyList
}

func (adjacencyList *AdjacencyList) RemoveVertex(u string) *AdjacencyList {
	if adjacencyList.graph[u] == nil {
		return adjacencyList
	}

	for vertex, list := range adjacencyList.graph {
		newList := []Vertex{}
		for _, neighbor := range list {
			if neighbor.Value != u {
				newList = append(newList, neighbor)
			}
		}
		adjacencyList.graph[vertex] = newList
	}

	delete(adjacencyList.graph, u)

	return adjacencyList
}

func (adjacencyList *AdjacencyList) AddEdge(u string, v string, options EdgeOptions) *AdjacencyList {
	if adjacencyList.graph[u] == nil || adjacencyList.graph[v] == nil {
		return adjacencyList
	}

	adjacencyList.graph[u] = append(adjacencyList.graph[u], Vertex{Value: v, Weight: options.Weight})

	if !options.Directed {
		adjacencyList.graph[v] = append(adjacencyList.graph[v], Vertex{Value: u, Weight: options.Weight})
	}

	return adjacencyList
}

func (adjacencyList *AdjacencyList) RemoveEdge(u string, v string, options EdgeOptions) *AdjacencyList {
	if adjacencyList.graph[u] == nil || adjacencyList.graph[v] == nil {
		return adjacencyList
	}

	newAdjacencyListU := []Vertex{}
	for _, neighbor := range adjacencyList.graph[u] {
		if neighbor.Value != v {
			newAdjacencyListU = append(newAdjacencyListU, neighbor)
		}
	}
	adjacencyList.graph[u] = newAdjacencyListU

	if options.Directed {
		return adjacencyList
	}

	newAdjacencyListV := []Vertex{}
	for _, neighbor := range adjacencyList.graph[v] {
		if neighbor.Value != u {
			newAdjacencyListV = append(newAdjacencyListV, neighbor)
		}
	}
	adjacencyList.graph[v] = newAdjacencyListV

	return adjacencyList
}

func (adjacencyList *AdjacencyList) GetNeighbors(u string) []Vertex {
	return adjacencyList.graph[u]
}
